keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentConstraintRunner.kt

```diff
@@ -0,0 +1,250 @@
+package com.netflix.spinnaker.keel.actuation
+
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.anyStateful
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.constraints.ConstraintEvaluator
+import com.netflix.spinnaker.keel.api.constraints.ConstraintStatus.PENDING
+import com.netflix.spinnaker.keel.api.constraints.StatefulConstraintEvaluator
+import com.netflix.spinnaker.keel.core.api.EnvironmentArtifactVetoes
+import com.netflix.spinnaker.keel.core.comparator
+import com.netflix.spinnaker.keel.persistence.KeelRepository
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+
+/**
+ * This class is responsible for running constraints and queueing them for approval.
+ * Approval into an environment is the responsibility of the [EnvironmentPromotionChecker].
+ */
+@Component
+class EnvironmentConstraintRunner(
+  private val repository: KeelRepository,
+  private val constraints: List<ConstraintEvaluator<*>>,
+  private val publisher: ApplicationEventPublisher
+) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  private val implicitConstraints: List<ConstraintEvaluator<*>> = constraints.filter { it.isImplicit() }
+  private val explicitConstraints: List<ConstraintEvaluator<*>> = constraints - implicitConstraints
+
+  // constraints that are only run if they are defined in a delivery config
+  private val statefulEvaluators: List<ConstraintEvaluator<*>> = explicitConstraints
+    .filterIsInstance<StatefulConstraintEvaluator<*, *>>()
+  private val statelessEvaluators = explicitConstraints - statefulEvaluators
+
+  // constraints that run for every environment in a delivery config but aren't shown to the user.
+  private val implicitStatefulEvaluators: List<ConstraintEvaluator<*>> = implicitConstraints
+    .filterIsInstance<StatefulConstraintEvaluator<*, *>>()
+  private val implicitStatelessEvaluators: List<ConstraintEvaluator<*>> = implicitConstraints - implicitStatefulEvaluators
+
+  /**
+   * Checks the environment and determines the version that should be approved,
+   * or null if there is no version that passes the constraints for an env + artifact combo.
+   * Queues that version for approval if it exists.
```

Comment:
It will queue the most recent one. Basically we find the artifact that we think should be approved for the environment, and queue that. then we follow up on in-progress stateful constraints in `handleOlderPendingVersions` because we wouldn't get to them in this function and we need to time them out somehow.

this function should approve at most one version

Reply:
> this function should approve at most one version

Reading the code, it looks to me like the`handleOlderPendingVersions` function queues for approval all of the members of the `pendingVersionsToCheck` set that meet constraints. What am I missing?

Reply:
Oops sorry :( i got confused

`fun checkConstraints()` approves at most one version
`fun handleOlderPendingVersions()` could approve an older version if it was already in progress

the function which calls both of them (which you commented on here) `fun checkEnvironment()` could approve lots of versions